### PR TITLE
Add ViewComponent::Deprecation

### PR DESCRIPTION
**Pull request by elia**
### Summary

This should allow consumers to see deprecation warnings that correctly mention the next version of VC.

Silencing is already a feature of AS::Deprecation

Closes #586 and closes #1273 (this was built starting from those PRs).

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -22,3 +23,7 @@ Layout/DotPosition:
 Layout/LineLength:
   Max: 120
   Enabled: true
+
+Style/FrozenStringLiteralComment:
+  Exclude:
+    - test/sandbox/app/components/jbuilder_component.json.jbuilder # not yet supported inside jbuilder templates
```
Comment (Spone) [inline .rubocop.yml:29]: What about supporting any future .jsbuilder file as well?

```suggestion
    - **/*.jbuilder # not yet supported inside jbuilder templates
```

File: .rubocop.yml [hunk 2]
```diff
@@ -9,6 +9,7 @@ AllCops:
     - gemfiles/**/*
     - test/tmp/**/*
     - vendor/bundle/**/*
+  NewCops: disable
```
Comment (Spone) [inline .rubocop.yml:12]: I'm wondering if it's the right move.

File: lib/view_component/deprecation.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+require "active_support/deprecation"
+
+module ViewComponent
+  Deprecation = ActiveSupport::Deprecation.new("3.0", "ViewComponent")
```
Comment (Spone) [inline lib/view_component/deprecation.rb:6]: We will need to bump the version here when releasing v3.0, right? If yes, we could add this step to `script/release`.

File: .rubocop.yml (hunk 2 above)
Comment (elia) [reply .rubocop.yml:12]: Moving this to a different PR

File: lib/view_component/polymorphic_slots.rb [hunk 4]
```diff
@@ -25,23 +25,26 @@ def renders_many(slot_name, callable = nil)
       end
 
       def register_polymorphic_slot(slot_name, types, collection:)
+        unless types.empty?
```
Comment (Spone) [inline lib/view_component/polymorphic_slots.rb:28]: This change feels unrelated, is it aiming at solving a warning?

Comment (elia) [reply lib/view_component/polymorphic_slots.rb:28]: Yes, the warning message is reported in the commit, but basically the getter method was redefined for each type.
The `unless types.empty?` guard is there just to keep the exact same behavior, i.e. if `types` was empty the getter is not defined.

**Review by Spone — APPROVED** [review]

File: lib/view_component.rb [hunk 5]
```diff
@@ -21,7 +22,7 @@ module ViewComponent
 
 # :nocov:
 if defined?(ViewComponent::Engine)
-  ActiveSupport::Deprecation.warn(
+  ViewComponent::Deprecation.warn(
     "This manually engine loading is deprecated and will be removed in v3.0.0. " \
```
Comment (camertron) [inline lib/view_component.rb:26]:
```suggestion
    "Manually loading the engine is deprecated and will be removed in v3.0.0. " \
```

**Review by camertron — APPROVED** [review]

**Comment by maxbeizer** [comment]
👋 thanks for picking up what I started and never got back around to 🙇 💖
